File: AndroidExpert/Dagger2App/app/src/main/java/com/algokelvin/dagger2/di/AppComponent.java
package com.algokelvin.dagger2.di;

import android.app.Application;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;
import dagger.android.AndroidInjector;
import dagger.android.DaggerApplication;
import dagger.android.support.AndroidSupportInjectionModule;

@Singleton
@Component(modules = {
        AndroidSupportInjectionModule.class,
        AppModule.class,
        ActivityModule.class
})
interface AppComponent extends AndroidInjector<DaggerApplication> {
    void inject(AppDagger appDagger);

    @Component.Builder
    interface Builder {
        @BindsInstance
        Builder application(Application application);
        AppComponent build();
    }
}

File: FeaturingAndroid/MovieTvPart2/app/src/main/java/algokelvin/app/movietvclient/presentation/di/core/AppComponent.kt
package algokelvin.app.movietvclient.presentation.di.core

import algokelvin.app.movietvclient.presentation.di.artist.ArtistSubComponent
import algokelvin.app.movietvclient.presentation.di.movie.MovieSubComponent
import algokelvin.app.movietvclient.presentation.di.tvShow.TvShowSubComponent
import dagger.Component
import javax.inject.Singleton

@Singleton
@Component(modules = [
    AppModule::class,
    CacheDataModule::class,
    DatabaseModule::class,
    LocalDataModule::class,
    NetModule::class,
    RemoteDataModule::class,
    RepositoryModule::class,
    UseCaseModule::class
])
interface AppComponent {

    fun movieSubComponent(): MovieSubComponent.Factory
    fun tvShowSubComponent(): TvShowSubComponent.Factory
    fun artistSubComponent(): ArtistSubComponent.Factory

}